File: solutions/javascript/sieve/1/sieve.js
export const primes = (limit) => {
  if (limit < 2) return [];
  
  // Generate vector of all candidate primes;
  let candidates = new Array(limit - 1)
    .fill(0)
    .map((e,i)=> i + 2);

  // Create map to isPrime boolean with all elements set true
  let isPrime = candidates.reduce((acc, e) => acc.set(e, true), new Map());

  // Execute Sieve
  for (let candidate of candidates) {
    if (!isPrime.get(candidate)) continue;

    for (let i = 2; i * candidate <= limit; i++) {
      isPrime.set(candidate * i, false);
    }
  }

  return candidates.filter(candidate => isPrime.get(candidate));
};

File: solutions/python/sieve/1/sieve.py
def calculate_primes(limit: int = 1_000_000) -> list[int]:
    if limit < 2:
        return []
    
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False

    for i in range(2, int(limit ** 0.5) + 1):
        if sieve[i]:
            for j in range(i * i, limit + 1, i):
                sieve[j] = False

    return [idx for idx, is_prime in enumerate(sieve) if is_prime]

def primes(limit: int) -> list[int]:
    return calculate_primes(limit)
